File: course-schedule-iv.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        indegree = [0]*numCourses
        childs = defaultdict(list)
        for p, ch in prerequisites:
            childs[p].append(ch)
            indegree[ch] += 1
        dq = deque()
        for i, d in enumerate(indegree):
            if d == 0:
                dq.append(i)
        parents = defaultdict(set)
        while dq:
            v = dq.popleft()
            for ch in childs[v]:
                parents[ch].update(parents[v])
                parents[ch].add(v)
                indegree[ch] -= 1
                if indegree[ch] == 0:
                    dq.append(ch)
        ans = []
        for p, ch in queries:
            ans.append(p in parents[ch])
        return ans

File: course-schedule-iv.cpp
int dis[100][100];
class Solution {
public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        for(int i = 0; i < numCourses; ++i){
            fill(dis[i], dis[i] + numCourses, 1e5);
        }

        for(auto &edge: prerequisites){
            dis[edge[0]][edge[1]] = 1;
        }

        for(int i = 0; i < numCourses; ++i){
            for(int j = 0; j < numCourses; ++j){
                for(int k = 0; k < numCourses; ++k){
                    dis[j][k] = min(dis[j][k], dis[j][i] + dis[i][k]);
                }
            }
        }
        vector<bool> ans;
        for(auto &q: queries){
            ans.push_back(dis[q[0]][q[1]] < 100000);
        }
        return ans;
    }
};